test/unit/attachment_data_visibility_test.rb

```diff
@@ -0,0 +1,643 @@
+    let(:attachment_data) { attachment.attachment_data }
+
+    before do
+      attachable.attachments << attachment
+      VirusScanHelpers.simulate_virus_scan(attachment_data.file)
+    end
+
+    context 'on a draft edition' do
+      let(:edition) { create(:news_article, organisations: [organisation]) }
+      let(:attachable) { edition }
+
+      it 'is not deleted' do
+        refute attachment_data.reload.deleted?
+      end
+
+      it 'is draft' do
+        assert attachment_data.reload.draft?
+      end
+
+      it 'is not accessible to anonymous user' do
+        refute attachment_data.reload.accessible_to?(anonymous_user)
+      end
+
+      it 'is accessible to user in same organisation' do
+        assert attachment_data.reload.accessible_to?(user_in_same_organisation)
+      end
+
+      it 'is accessible to user in another organisation' do
+        assert attachment_data.reload.accessible_to?(user_in_another_organisation)
+      end
+
+      it 'is not unpublished' do
+        refute attachment_data.reload.unpublished?
+      end
+
+      it 'has no unpublished edition' do
+        assert_nil attachment_data.reload.unpublished_edition
+      end
+
+      it 'is not replaced' do
+        refute attachment_data.reload.replaced?
+      end
+
+      context 'edition is access-limited' do
+        before do
+          edition.access_limited = true
+          edition.save!
+        end
+
+        it 'is not accessible to anonymous user' do
+          refute attachment_data.reload.accessible_to?(anonymous_user)
+        end
+
+        it 'is accessible to user in same organisation' do
+          assert attachment_data.reload.accessible_to?(user_in_same_organisation)
+        end
+
+        it 'is not accessible to user in another organisation' do
+          refute attachment_data.reload.accessible_to?(user_in_another_organisation)
+        end
+      end
+
+      context 'when attachment is deleted' do
+        before do
+          attachment.destroy!
+        end
+
+        it 'is deleted' do
+          assert attachment_data.reload.deleted?
+        end
+      end
+
+      context 'when attachment is replaced' do
+        before do
+          attributes = attributes_for(:attachment_data)
+          attributes[:to_replace_id] = attachment_data.id
+          attachment.update_attributes!(attachment_data_attributes: attributes)
+        end
+
+        it 'is not deleted' do
+          refute attachment_data.reload.deleted?
+        end
+
+        it 'is draft' do
+          assert attachment_data.reload.draft?
+        end
+
+        it 'is replaced' do
+          assert attachment_data.reload.replaced?
+        end
+
+        context 'when edition is published' do
+          before do
+            edition.major_change_published_at = Time.zone.now
+            edition.force_publish!
+          end
+
+          it 'is not deleted' do
+            refute attachment_data.reload.deleted?
+          end
+
+          it 'is draft' do
```

Comment:
We discussed adding some more text to explain why this is draft - because it's the original `attachment_data` that's been replaced and has never been published.

Reply:
We talked about adding something like `refute attachment.reload.attachment_data.draft?` so that we're illustrating the difference between the original and new `AttachmentData` instances.

Reply:
I've added the following paragraph to the commit note:

> Note that all the assertions are against the *original* `AttachmentData`, rather than the replacement `AttachmentData`. This explains why, for example, when an attachment is added to a draft edition, the attachment is replaced, and the edition is published, then the original `AttachmentData` is still draft, because it was never published; whereas the replacement `AttachmentData` is not draft.

Reply:
It turns out I forgot to push up the changes to the commit note before I merged the PR, so the above comment will have to suffice - sorry!